### ISSUE TITLE
six.reraise in traceback output

### ISSUE THREAD

**Issue by The-Compiler** [pytest-dev/pytest#5143]
I unfortunately don't have a ready reproducer for this - but I just noticed this in the pytest output of a failing test (Python 3.7, pytest 4.4.1):

```
________________ ERROR at teardown of test_opening_a_quickmark _________________
tp = <class 'AssertionError'>, value = None, tb = None
    def reraise(tp, value, tb=None):
        try:
            if value is None:
                value = tp()
            if value.__traceback__ is not tb:
                raise value.with_traceback(tb)
>           raise value
.tox/py37-pyqt511/lib/python3.7/site-packages/six.py:693: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
.tox/py37-pyqt511/lib/python3.7/site-packages/six.py:693: in reraise
    raise value
.tox/py37-pyqt511/lib/python3.7/site-packages/six.py:693: in reraise
    raise value
tests/end2end/fixtures/quteprocess.py:907: in quteproc_process
    proc.terminate()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
self = <end2end.fixtures.quteprocess.QuteProc object at 0x7f34c690cf78>
    def terminate(self):
        """Clean up and shut down the process."""
        if not self.is_running():
            return
    
        if quteutils.is_windows:
            self.proc.kill()
        else:
            self.proc.terminate()
    
        ok = self.proc.waitForFinished()
>       assert ok
E       assert False
tests/end2end/fixtures/testprocess.py:313: AssertionError
```

There's a lot of `six.reraise` fluff in there, without me using it in my own code. I suppose that's pytest using it internally somehow?

Would be glad [on some pointers](https://www.xkcd.com/138/) how to track this down.

Full traceback with `--full-trace`:

<details>

```
cls = <class '_pytest.runner.CallInfo'>, func = <function call_runtest_hook.<locals>.<lambda> at 0x7fd6a30cd620>, when = 'teardown'
reraise = (<class '_pytest.outcomes.Exit'>, <class 'KeyboardInterrupt'>)

    @classmethod
    def from_call(cls, func, when, reraise=None):
        #: context of invocation: one of "setup", "call",
        #: "teardown", "memocollect"
        start = time()
        excinfo = None
        try:
>           result = func()

.tox/py37-pyqt512/lib/python3.7/site-packages/_pytest/runner.py:226: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>       lambda: ihook(item=item, **kwds), when=when, reraise=reraise
    )

.tox/py37-pyqt512/lib/python3.7/site-packages/_pytest/runner.py:198: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <_HookCaller 'pytest_runtest_teardown'>, args = (), kwargs = {'item': <Function test_opening_a_quickmark>, 'nextitem': None}, notincall = set()

    def __call__(self, *args, **kwargs):
        if args:
            raise TypeError("hook calling supports only keyword arguments")
        assert not self.is_historic()
        if self.spec and self.spec.argnames:
            notincall = (
                set(self.spec.argnames) - set(["__multicall__"]) - set(kwargs.keys())
            )
            if notincall:
                warnings.warn(
                    "Argument(s) {} which are declared in the hookspec "
                    "can not be found in this hook call".format(tuple(notincall)),
                    stacklevel=2,
                )
>       return self._hookexec(self, self.get_hookimpls(), kwargs)

.tox/py37-pyqt512/lib/python3.7/site-packages/pluggy/hooks.py:289: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <_pytest.config.PytestPluginManager object at 0x7fd6b78e2358>, hook = <_HookCaller 'pytest_runtest_teardown'>
methods = [<HookImpl plugin_name='runner', plugin=<module '_pytest.runner' from '/home/florian/proj/qutebrowser/git/.tox/py37-py...quickmark>>>, <HookImpl plugin_name='logging-plugin', plugin=<_pytest.logging.LoggingPlugin object at 0x7fd6a3a9a4e0>>]
kwargs = {'item': <Function test_opening_a_quickmark>, 'nextitem': None}

    def _hookexec(self, hook, methods, kwargs):
        # called from all hookcaller instances.
        # enable_tracing will set its own wrapping function at self._inner_hookexec
>       return self._inner_hookexec(hook, methods, kwargs)

.tox/py37-pyqt512/lib/python3.7/site-packages/pluggy/manager.py:68: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

hook = <_HookCaller 'pytest_runtest_teardown'>
methods = [<HookImpl plugin_name='runner', plugin=<module '_pytest.runner' from '/home/florian/proj/qutebrowser/git/.tox/py37-py...quickmark>>>, <HookImpl plugin_name='logging-plugin', plugin=<_pytest.logging.LoggingPlugin object at 0x7fd6a3a9a4e0>>]
kwargs = {'item': <Function test_opening_a_quickmark>, 'nextitem': None}

    self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
        methods,
        kwargs,
>       firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
    )

.tox/py37-pyqt512/lib/python3.7/site-packages/pluggy/manager.py:62: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

item = <Function test_opening_a_quickmark>, nextitem = None

    def pytest_runtest_teardown(item, nextitem):
        _update_current_test_var(item, "teardown")
>       item.session._setupstate.teardown_exact(item, nextitem)

.tox/py37-pyqt512/lib/python3.7/site-packages/_pytest/runner.py:137: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <_pytest.runner.SetupState object at 0x7fd6a3aa5780>, item = <Function test_opening_a_quickmark>, nextitem = None

    def teardown_exact(self, item, nextitem):
        needed_collectors = nextitem and nextitem.listchain() or []
>       self._teardown_towards(needed_collectors)

.tox/py37-pyqt512/lib/python3.7/site-packages/_pytest/runner.py:332: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <_pytest.runner.SetupState object at 0x7fd6a3aa5780>, needed_collectors = []

    def _teardown_towards(self, needed_collectors):
        exc = None
        while self.stack:
            if self.stack == needed_collectors[: len(self.stack)]:
                break
            try:
                self._pop_and_teardown()
            except TEST_OUTCOME:
                # XXX Only first exception will be seen by user,
                #     ideally all should be reported.
                if exc is None:
                    exc = sys.exc_info()
        if exc:
>           six.reraise(*exc)

.tox/py37-pyqt512/lib/python3.7/site-packages/_pytest/runner.py:347: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

tp = <class 'AssertionError'>, value = None, tb = None

    def reraise(tp, value, tb=None):
        try:
            if value is None:
                value = tp()
            if value.__traceback__ is not tb:
                raise value.with_traceback(tb)
>           raise value

.tox/py37-pyqt512/lib/python3.7/site-packages/six.py:693: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <_pytest.runner.SetupState object at 0x7fd6a3aa5780>, needed_collectors = []

    def _teardown_towards(self, needed_collectors):
        exc = None
        while self.stack:
            if self.stack == needed_collectors[: len(self.stack)]:
                break
            try:
>               self._pop_and_teardown()

.tox/py37-pyqt512/lib/python3.7/site-packages/_pytest/runner.py:340: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <_pytest.runner.SetupState object at 0x7fd6a3aa5780>

    def _pop_and_teardown(self):
        colitem = self.stack.pop()
>       self._teardown_with_finalization(colitem)

.tox/py37-pyqt512/lib/python3.7/site-packages/_pytest/runner.py:297: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <_pytest.runner.SetupState object at 0x7fd6a3aa5780>, colitem = <Module tests/end2end/features/test_open_bdd.py>

    def _teardown_with_finalization(self, colitem):
>       self._callfinalizers(colitem)

.tox/py37-pyqt512/lib/python3.7/site-packages/_pytest/runner.py:315: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <_pytest.runner.SetupState object at 0x7fd6a3aa5780>, colitem = <Module tests/end2end/features/test_open_bdd.py>

    def _callfinalizers(self, colitem):
        finalizers = self._finalizers.pop(colitem, None)
        exc = None
        while finalizers:
            fin = finalizers.pop()
            try:
                fin()
            except TEST_OUTCOME:
                # XXX Only first exception will be seen by user,
                #     ideally all should be reported.
                if exc is None:
                    exc = sys.exc_info()
        if exc:
>           six.reraise(*exc)

.tox/py37-pyqt512/lib/python3.7/site-packages/_pytest/runner.py:312: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

tp = <class 'AssertionError'>, value = None, tb = None

    def reraise(tp, value, tb=None):
        try:
            if value is None:
                value = tp()
            if value.__traceback__ is not tb:
                raise value.with_traceback(tb)
>           raise value

.tox/py37-pyqt512/lib/python3.7/site-packages/six.py:693: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <_pytest.runner.SetupState object at 0x7fd6a3aa5780>, colitem = <Module tests/end2end/features/test_open_bdd.py>

    def _callfinalizers(self, colitem):
        finalizers = self._finalizers.pop(colitem, None)
        exc = None
        while finalizers:
            fin = finalizers.pop()
            try:
>               fin()

.tox/py37-pyqt512/lib/python3.7/site-packages/_pytest/runner.py:305: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <FixtureDef argname='quteproc_process' scope='module' baseid='tests/end2end'>, request = <SubRequest 'quteproc_process' for <Function test_opening_a_quickmark>>

    def finish(self, request):
        exceptions = []
        try:
            while self._finalizers:
                try:
                    func = self._finalizers.pop()
                    func()
                except:  # noqa
                    exceptions.append(sys.exc_info())
            if exceptions:
                e = exceptions[0]
                del exceptions  # ensure we don't keep all frames alive because of the traceback
>               six.reraise(*e)

.tox/py37-pyqt512/lib/python3.7/site-packages/_pytest/fixtures.py:857: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

tp = <class 'AssertionError'>, value = None, tb = None

    def reraise(tp, value, tb=None):
        try:
            if value is None:
                value = tp()
            if value.__traceback__ is not tb:
                raise value.with_traceback(tb)
>           raise value

.tox/py37-pyqt512/lib/python3.7/site-packages/six.py:693: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <FixtureDef argname='quteproc_process' scope='module' baseid='tests/end2end'>, request = <SubRequest 'quteproc_process' for <Function test_opening_a_quickmark>>

    def finish(self, request):
        exceptions = []
        try:
            while self._finalizers:
                try:
                    func = self._finalizers.pop()
>                   func()

.tox/py37-pyqt512/lib/python3.7/site-packages/_pytest/fixtures.py:851: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

fixturefunc = <function quteproc_process at 0x7fd6a3aecbf8>, it = <generator object quteproc_process at 0x7fd6a307d7c8>

    def _teardown_yield_fixture(fixturefunc, it):
        """Executes the teardown of a fixture function by advancing the iterator after the
        yield and ensure the iteration ends (if not it means there is more than one yield in the function)"""
        try:
>           next(it)

.tox/py37-pyqt512/lib/python3.7/site-packages/_pytest/fixtures.py:802: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

qapp = <PyQt5.QtWidgets.QApplication object at 0x7fd6a3197a68>, server = <end2end.fixtures.webserver.WebserverProcess object at 0x7fd6a3197af8>
request = <SubRequest 'quteproc_process' for <Function test_opening_a_quickmark>>

    @pytest.fixture(scope='module')
    def quteproc_process(qapp, server, request):
        """Fixture for qutebrowser process which is started once per file."""
        # Passing request so it has an initial config
        proc = QuteProc(request)
        proc.start()
        yield proc
>       proc.terminate()

tests/end2end/fixtures/quteprocess.py:907: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <end2end.fixtures.quteprocess.QuteProc object at 0x7fd6a3acc9d8>

    def terminate(self):
        """Clean up and shut down the process."""
        if not self.is_running():
            return
    
        if quteutils.is_windows:
            self.proc.kill()
        else:
            self.proc.terminate()
    
        ok = self.proc.waitForFinished()
        ok = False
>       assert ok
E       assert False

tests/end2end/fixtures/testprocess.py:314: AssertionError
```
</details>

**Comment by blueyed** [comment]
How does `--tb=native` look like?

**Comment by The-Compiler** [comment]
Looks like that somehow gets cut as well?

```
Traceback (most recent call last):
  File "/home/florian/proj/qutebrowser/git/.tox/py37-pyqt512/lib/python3.7/site-packages/six.py", line 693, in reraise
    raise value
  File "/home/florian/proj/qutebrowser/git/.tox/py37-pyqt512/lib/python3.7/site-packages/_pytest/runner.py", line 340, in _teardown_towards
    self._pop_and_teardown()
  File "/home/florian/proj/qutebrowser/git/.tox/py37-pyqt512/lib/python3.7/site-packages/_pytest/runner.py", line 297, in _pop_and_teardown
    self._teardown_with_finalization(colitem)
  File "/home/florian/proj/qutebrowser/git/.tox/py37-pyqt512/lib/python3.7/site-packages/_pytest/runner.py", line 315, in _teardown_with_finalization
    self._callfinalizers(colitem)
  File "/home/florian/proj/qutebrowser/git/.tox/py37-pyqt512/lib/python3.7/site-packages/_pytest/runner.py", line 312, in _callfinalizers
    six.reraise(*exc)
  File "/home/florian/proj/qutebrowser/git/.tox/py37-pyqt512/lib/python3.7/site-packages/six.py", line 693, in reraise
    raise value
  File "/home/florian/proj/qutebrowser/git/.tox/py37-pyqt512/lib/python3.7/site-packages/_pytest/runner.py", line 305, in _callfinalizers
    fin()
  File "/home/florian/proj/qutebrowser/git/.tox/py37-pyqt512/lib/python3.7/site-packages/_pytest/fixtures.py", line 857, in finish
    six.reraise(*e)
  File "/home/florian/proj/qutebrowser/git/.tox/py37-pyqt512/lib/python3.7/site-packages/six.py", line 693, in reraise
    raise value
  File "/home/florian/proj/qutebrowser/git/.tox/py37-pyqt512/lib/python3.7/site-packages/_pytest/fixtures.py", line 851, in finish
    func()
  File "/home/florian/proj/qutebrowser/git/.tox/py37-pyqt512/lib/python3.7/site-packages/_pytest/fixtures.py", line 802, in _teardown_yield_fixture
    next(it)
  File "/home/florian/proj/qutebrowser/git/tests/end2end/fixtures/quteprocess.py", line 907, in quteproc_process
    proc.terminate()
  File "/home/florian/proj/qutebrowser/git/tests/end2end/fixtures/testprocess.py", line 314, in terminate
    assert ok
AssertionError: assert False
```

I tried `--tb=native` with `--full-trace` but then it just seems to be ignored.

**Comment by blueyed** [comment]
Looks related to this XXX: https://github.com/pytest-dev/pytest/blob/e87d3d70e2da2663ee619ec60e878be928d49506/src/_pytest/runner.py#L342-L343

**Comment by blueyed** [comment]
A minimal test case would certainly help to fix this in the future.
Looks a bit like you might have two exeptions there, and then the first one gets re-raised, and this causes it to be cut off then maybe?

**Comment by Zac-HD** [comment]
Hmm.  TBH I'd just leave this alone for a few weeks more, and we'll drop `six` with Python 2 in pytest 5.0.

**Comment by nicoddemus** [comment]
We no longer use `six` on `master` so this should be fixed in 5.0. 👍
